README.md
![alt tag](https://raw.githubusercontent.com/lateralblast/clq/master/clq.jpg)

CLQ
===

Command Line Quiz

Introduction
------------

A simple shell and ruby script that converts a formatted CSV file into a multiple choice quiz.

Shell script: clq.sh

Ruby script: clq.rb

Supported Operating Systems
---------------------------

Any operating system with sh/bash or ruby

License
-------

This software is licensed as CC-BA (Creative Commons By Attrbution)

http://creativecommons.org/licenses/by/4.0/legalcode

Features
--------

Some of the features:

- Doesn't wait for enter key to be pressed increasing speed users can go through questions
- Responds with correct answer, coloring the text green if the answer was correct or red if it wasn't
- Has support for multiple answer questions (i.e. choose to or more correct answers)
- Keeps track of correct and wrong answers providing a tally at the end
- Allows quiting at anytime with 'q'
- Ability to ask questions in random order
- Ability to mix choices from other questions
  - Handy for learning vocabulary where you want to mix things up and avoid pattern recognition
  - Not necessarily handy for quiz with disparate questions types as the outliers are obvious


Shell Script Usage Information
------------------------------

Get help:

```
$ ./clq.sh -h

Usage: ./clq.sh -[h|V|q] [quiz]

-h:        Print usage
-V:        Print version
-q [quiz]: Quiz (ask questions in sequential order)
-r [quiz]: Quiz (ask questions in random order)
```

List available quizes:

```
$ ./clq.sh -l
Available quizes:
example
```

Do the multiple choice quiz example:

```
$ ./clq.sh -q example
```

Ask questions in random order:

```
$ ./clq.sh -r example
```

Ruby Script Usage Information
-----------------------------

Get help:

```
$ ./clq.rb -h

Usage: ./clq.rb

"--list",     "-l"  List quizes
"--random",   "-r"  Randomise quizes
"--mix",      "-m"  Mix choices between questions
"--quiz",     "-q"  Quiz
"--help",     "-h"  Print help information
"--version",  "-V"  Print version information
```

List available quizes:

```
$ ./clq.rb -l
Available quizes:
example
```

Do the multiple choice quiz example:

```
$ ./clq.rb -q example
```

Ask questions in random order:

```
$ ./clq.rb -q example -r
```

Ask questions in random order and mix choices between questions

```
$ ./clq.rb -q example -r -m
```

Question File Information
-------------------------

This script uses a CSV file to generate a set of multi-choice questions.

It uses the pipe "|" symbol as a delimiter.

The header, and format of the CSV file is as follows:

```
Question|Answer|A|B|C|D|E
```

The header must be included as the first line of the CSV file.

An example of a file:

```
Question|Answer|A|B|C|D|E
Amazon Glacier is designed for: (Choose 2 answers)|B,C|active database storage.|infrequently accessed data.|data archives.|frequently accessed data.|cached session data.
Your web application front end consists of multiple EC2 instances behind an Elastic Load Balancer. You configured ELB to perform health checks on these EC2 instances. If an instance fails to pass health checks, which statement will be true?|C|The instance is replaced automatically by the ELB.|The instance gets terminated automatically by the ELB.|The ELB stops sending traffic to the instance that failed its health check.|The instance gets quarantined by the ELB for root cause analysis.|
You are building a system to distribute confidential training videos to employees. Using CloudFront, what method could be used to serve content that is stored in S3, but not publically accessible from S3 directly?|A|Create an Origin Access Identity (OAI) for CloudFront and grant access to the objects in your S3 bucket to that OAI.|Add the CloudFront account security group “amazon-cf/amazon-cf-sg” to the appropriate S3 bucket policy.|Create an Identity and Access Management (IAM) User for CloudFront and grant access to the objects in your S3 bucket to that IAM User.|Create a S3 bucket policy that lists the CloudFront distribution ID as the Principal and the target bucket as the Amazon Resource Name (ARN).|
```

From this example we can see the first question allows two answers, B and C.
In this case the user would enter 'bc' when asked for an answer.

The second question has C as the correct answer

Examples
--------

List available quizes:

```
$ ./clq.sh -l
Available quizes:
example
```

Do the multiple choice quiz example in random order and mix choices amongst questions:

```
$ ./clq.rb -q example -m -r

Your web application front end consists of multiple EC2 instances behind an
Elastic Load Balancer. You configured ELB to perform health checks on these EC2
instances. If an instance fails to pass health checks, which statement will be
true?

A: The instance gets quarantined by the ELB for root cause analysis.

B: Create an Origin Access Identity (OAI) for CloudFront and grant access to the
   objects in your S3 bucket to that OAI.

C: The instance gets terminated automatically by the ELB.

D: The ELB stops sending traffic to the instance that failed its health check.


Answer? d

D: The ELB stops sending traffic to the instance that failed its health check.


You are building a system to distribute confidential training videos to
employees. Using CloudFront, what method could be used to serve content that is
stored in S3, but not publically accessible from S3 directly?

A: cached session data.

B: active database storage.

C: Create an Origin Access Identity (OAI) for CloudFront and grant access to the
   objects in your S3 bucket to that OAI.

D: The instance gets quarantined by the ELB for root cause analysis.


Answer? c

C: Create an Origin Access Identity (OAI) for CloudFront and grant access to the
objects in your S3 bucket to that OAI.


Amazon Glacier is designed for: (Choose 2 answers)

A: Create an Identity and Access Management (IAM) User for CloudFront and grant
   access to the objects in your S3 bucket to that IAM User.

B: infrequently accessed data.

C: Create a S3 bucket policy that lists the CloudFront distribution ID as the
   Principal and the target bucket as the Amazon Resource Name (ARN).

D: data archives.

E: Create an Origin Access Identity (OAI) for CloudFront and grant access to the
   objects in your S3 bucket to that OAI.


Answer? bd

B: infrequently accessed data.
D: data archives.



Results:

Questions: 3
Correct:   3
Wrong:     0
Percent:   100.0%
```

changelog
# 0.0.1 Sat  4 Feb 2017 12:40:02 AEDT
# Initial working version
# 0.0.2 Sat  4 Feb 2017 19:55:11 AEDT
# Added option to randomise order of questions

clq.sh
#!/bin/sh

# Name:         clq (Command Line Quiz)
# Version:      0.0.3
# Release:      1
# License:      CC-BA (Creative Commons By Attribution)
#               http://creativecommons.org/licenses/by/4.0/legalcode
# Group:        System
# Source:       N/A
# URL:          http://lateralblast.com.au/
# Distribution: UNIX
# Vendor:       UNIX
# Packager:     Richard Spindler <richard@lateralblast.com.au>
# Description:  A POC shell script to turn a formatted csv file into multiple choice quiz

# Get command line args

args=$@

# Set some defaults

quiz_dir="quizes"
text_green="\e[0;32m"
text_red="\e[0;31m"
text_white="\e[0;37m"

# Get the path the script starts from

start_path=`pwd`

# Get the version of the script from the script itself

script_version=`cd $start_path ; cat $0 | grep '^# Version' |awk '{print $3}'`

# Get the script name

script_name=`cd $start_path ; cat $0 | grep '^# Name' |awk '{print $3}'`

# print_usage
#
# If given a -h or no valid switch print usage information

print_usage () {
  echo ""
  echo "Usage: $0 -[h|V|q] [quiz]"
  echo ""
  echo "-h:        Print usage"
  echo "-V:        Print version"
  echo "-q [quiz]: Quiz (ask questions in sequential order)"
  echo "-r [quiz]: Quiz (ask questions in random order)"
  echo ""
}

# If given no command line arguments print usage information

if [ `expr "$args" : "\-"` != 1 ]; then
  print_usage
fi

# List quizes

list_quizes () {
  echo "Available quizes:"
  ls $quiz_dir
}

# Get ASCII value for letter

ord() {
  LC_CTYPE=C printf '%d' "'$1"
}

# Print results

print_results () {
  no_questions=$1
  no_correct=$2
  no_wrong=$3
  percent=$(awk "BEGIN { pc=100*$no_correct/$no_questions; i=int(pc); print (pc-i<0.5)?i:i+1 }")
  echo ""
  echo ""
  echo "Results:"
  echo ""
  echo "Questions: $no_questions"
  echo "Correct:   $no_correct"
  echo "Wrong:     $no_wrong"
  echo "Percent:   $percent%"
  echo ""
}

# Handle quiz

handle_quiz () {
  quiz_file=$1
  random=$2
  printf $text_white
  if [ ! -f "$quiz_file" ]; then
    orig_file="$quiz_file"
    quiz_file="$quiz_dir/$orig_file"
    if [ ! -f "$quiz_file" ]; then
      echo "Cannot find quiz: $orig_file"
      exit
    fi
  fi 
  no_questions=0
  no_correct=0
  no_wrong=0
  header=`cat $quiz_file |grep "^Question"`
  header_a=`echo "$header" |cut -f3 -d'|'`
  header_b=`echo "$header" |cut -f4 -d'|'`
  header_c=`echo "$header" |cut -f5 -d'|'`
  header_d=`echo "$header" |cut -f6 -d'|'`
  header_e=`echo "$header" |cut -f7 -d'|'`
  if [ "$random" -eq 1 ]; then
    quiz_data=`cat $quiz_file |awk 'BEGIN{srand();} {printf "%06d %s\n", rand()*1000000, $0;}' | sort -n | cut -c8-`
  else
    quiz_data=`cat $quiz_file`
  fi
  while read -r line <&3; do
    if [ ! "$line" = "" ]; then
      question=`echo "$line" |cut -f1 -d'|'`
      if [ ! "$question" = "Question" ]; then
        echo ""
        echo "$question" |fmt -w 80
        echo ""
        correct=`echo "$line" |cut -f2 -d'|' |sed 's/[ ,\,]//g' |tr "[:upper:]" "[:lower:]" |grep -o . | sort |tr -d "\n"`
        answer=""
        for counter in $(seq 1 ${#correct}); do
          char=${correct:counter-1:1}
          ascii=$(ord $char)
          field=`expr $ascii - 94`
          char=`echo $char |tr "[:lower:]" "[:upper:]"`
          string=`echo "$line" |cut -f$field -d'|'`
          if [ "$answer" = "" ]; then
            answer="$char - $string"
          else
            answer="$answer $char - $string"
          fi
        done
        chars=`echo "$correct" |wc -c`
        chars=`expr $chars - 1`
        choice_a=`echo "$line" |cut -f3 -d'|'`
        if [ ! "$choice_a" = "" ]; then
          string="$header_a: $choice_a"
          echo "$string" |fmt -w 80
        fi
        choice_b=`echo "$line" |cut -f4 -d'|'`
        if [ ! "$choice_b" = "" ]; then
          string="$header_b: $choice_b"
          echo "$string" |fmt -w 80
        fi
        choice_c=`echo "$line" |cut -f5 -d'|'`
        if [ ! "$choice_c" = "" ]; then
          string="$header_c: $choice_c"
          echo "$string" |fmt -w 80
        fi
        choice_d=`echo "$line" |cut -f6 -d'|'`
        if [ ! "$choice_d" = "" ]; then
          string="$header_d: $choice_d"
          echo "$string" |fmt -w 80
        fi
        choice_e=`echo "$line" |cut -f7 -d'|'`
        if [ ! "$choice_e" = "" ]; then
          string="$header_e: $choice_e"
          echo "$string" |fmt -w 80
        fi
        echo ""
        read -p "Answer? " -n$chars response
        response=`echo "$response" | tr "[:upper:]" "[:lower:]" |grep -o . | sort |tr -d "\n"`
        if [ "$response" = "q" ]; then
          print_results $no_questions $no_correct $no_wrong
          exit
        fi
        if [ "$response" = "$correct" ]; then
          no_correct=`expr $no_correct + 1`
          text_colour=$text_green
        else
          no_wrong=`expr $no_wrong + 1`
          text_colour=$text_red
        fi
        no_questions=`expr $no_questions + 1`
        echo ""
        echo ""
        printf "$text_colour$answer$text_white"
        echo ""
      fi
    fi
  done 3<<< "$quiz_data"
  print_results $no_questions $no_correct $no_wrong
}

while getopts hlrVq: args; do
  case $args in
    l)
      list_quizes
      ;;
    r)
      quiz_file=$2
      random=1
      handle_quiz $quiz_file $random
      exit
      ;;
    q)
      quiz_file=$2
      random=0
      handle_quiz $quiz_file $random
      exit
      ;;
    h)
      print_usage
      ;;
    V)
      echo $script_version
      exit
      ;;
    *)
      print_usage
      exit
  esac
done



clq.rb
#!/usr/bin/env ruby

# Name:         clq (Command Line Quiz)
# Version:      0.0.9
# Release:      1
# License:      CC-BA (Creative Commons By Attribution)
#               http://creativecommons.org/licenses/by/4.0/legalcode
# Group:        System
# Source:       N/A
# URL:          http://lateralblast.com.au/
# Distribution: UNIX
# Vendor:       UNIX
# Packager:     Richard Spindler <richard@lateralblast.com.au>
# Description:  A POC ruby script to turn a formatted csv file into multiple choice quiz

require 'rubygems'
require 'io/console'

def install_gem(load_name,install_name)
  puts "Information:\tInstalling #{install_name}"
  %x[gem install #{install_name}]
  Gem.clear_paths
  require "#{load_name}"
end

begin
  require 'getopt/long'
rescue LoadError
  install_gem("getopt/long","getopt")
end
begin
  require 'smarter_csv'
rescue LoadError
  install_gem('smarter_csv','smarter_csv')
end
begin
  require 'colorize'
rescue LoadError
  install_gem('colorize','colorize')
end

class String
  def wrap_text(col = 80)
    gsub(/(.{1,#{col}})( +|$)\n?|(.{#{col}})/,"\\1\\3\n")
  end
  def wrap_text_with_indent(col = 80, indent = "   ")
    gsub(/(.{1,#{col}})( +|$)\n?|(.{#{col}})/,"\\1\\3\n#{indent}")
  end
end

include Getopt

# Set some defaults

$quiz_dir = "quizes"
$script   = $0

# Print script usage information

def print_usage()
  switches     = []
  long_switch  = ""
  short_switch = ""
  help_info    = ""
  puts ""
  puts "Usage: #{$script}"
  puts ""
  file_array  = IO.readlines $0
  option_list = file_array.grep(/\[ "--/)
  option_list.each do |line|
    if !line.match(/file_array/)
      help_info    = line.split(/# /)[1]
      switches     = line.split(/,/)
      long_switch  = switches[0].gsub(/\[/,"").gsub(/\s+/,"")
      short_switch = switches[1].gsub(/\s+/,"")
      if short_switch.match(/REQ|BOOL/)
        short_switch = ""
      end
      if long_switch.gsub(/\s+/,"").length < 7
        puts "#{long_switch},\t\t#{short_switch}\t#{help_info}"
      else
        puts "#{long_switch},\t#{short_switch}\t#{help_info}"
      end
    end
  end
  puts ""
  return
end

# Get version

def get_version()
  file_array = IO.readlines $0
  version    = file_array.grep(/^# Version/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  packager   = file_array.grep(/^# Packager/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  name       = file_array.grep(/^# Name/)[0].split(":")[1].gsub(/^\s+/,'').chomp
  return version,packager,name
end

# Print script version information

def print_version()
  (version,packager,name) = get_version()
  puts "#{name} v. #{version} #{packager}"
  return
end

# Get command line arguments
# Print help if specified none

if !ARGV[0]
  print_usage()
end

# List quizes

def list_quizes
  puts "Available quizes:"
  list = %x[ls #{$quiz_dir}]
  puts list
end

# Print results

def print_results(no_quest,no_right,no_wrong)
  if no_quest != 0
    percent = ( no_right.to_f / no_quest.to_f ) * 100
    percent = percent.round(1)
    puts ""
    puts ""
    puts "Results:"
    puts ""
    puts "Questions: "+no_quest.to_s
    puts "Correct:   "+no_right.to_s
    puts "Wrong:     "+no_wrong.to_s
    puts "Percent:   "+percent.to_s+"%"
  end
  puts ""
  return
end

# Handle quizes

def handle_quiz(quiz_file,random,mix)
  no_right = 0
  no_wrong = 0
  no_quest = 0
  if !File.exist?(quiz_file)
    test_file = $quiz_dir+"/"+quiz_file
    if !File.exist?(test_file)
      puts "Quiz "+quiz_file+" does not exist"
      exit
    else
      quiz_file = test_file
    end
  end
  quiz_data = SmarterCSV.process(quiz_file,:col_sep => "|")
  if random == true
    quiz_data = quiz_data.shuffle
  end
  if mix == true
    q_mix = []
    quiz_data.each do |key, value|
      [ 'a', 'b', 'c', 'd', 'e' ].each do |letter|
        q_mix.push(key[:"#{letter}"])
      end
    end
  end
  quiz_data.each do |key, value|
    r_correct = []
    answer    = ""
    text      = ""
    correct   = key[:answer].downcase.gsub(/,| /,"").chars.sort.join
    question  = key[:question]
    question  = question.wrap_text
    puts
    puts question
    puts
    if random == true
      counter = 0
      [ 'a', 'b', 'c', 'd', 'e' ].shuffle.each do |letter|
        if key[:"#{letter}"]
          choice  = "a".ord+counter
          choice  = choice.chr
          counter = counter + 1
          line    = ""
          test = 0
          line = ""
          text = ""
          if correct.match(/#{letter}/)
            r_correct.push(choice)
            if answer.length < 1
              answer = choice.upcase+": "+key[:"#{letter}"]
            else
              answer = answer+"\n"+choice.upcase+": "+key[:"#{letter}"]
            end
            line = choice.upcase.bold+": "+key[:"#{letter}"].wrap_text_with_indent
          else
            if mix == true
              text = q_mix.sample
              while test == 0
                test = 1
                text = q_mix.sample
                correct.split("").each do |sample|
                  if key[:"#{sample}"].match(/#{text}/) or key[:"#{letter}"].match(/#{text}/) or key[:"#{choice}"].match(/#{text}/)
                    test = 0
                  end
                end
              end
              line = choice.upcase.bold+": "+text.wrap_text_with_indent
            else 
              line = choice.upcase.bold+": "+key[:"#{letter}"].wrap_text_with_indent
            end
          end
          puts line
        end
      end
      correct = r_correct.join
    else
      [ 'a', 'b', 'c', 'd', 'e' ].each do |letter|
        test = 0
        line = ""
        text = ""
        if key[:"#{letter}"]
          if correct.match(/#{letter}/)
            if answer.length < 1
              answer = letter.upcase+": "+key[:"#{letter}"]
            else
              answer = answer+"\n"+letter.upcase+": "+key[:"#{letter}"]
            end
            line = letter.upcase.bold+": "+key[:"#{letter}"].wrap_text_with_indent
          else
            if mix == true
              text = q_mix.sample
              while test == 0
                test = 1
                text = q_mix.sample
                correct.split("").each do |sample|
                  if key[:"#{sample}"].match(/#{text}/) or key[:"#{letter}"].match(/#{text}/)
                    test = 0
                  end
                end
              end
              line = letter.upcase.bold+": "+text.wrap_text_with_indent
            else
              line = letter.upcase.bold+": "+key[:"#{letter}"].wrap_text_with_indent
            end
          end
          puts line
        end
      end
    end
    puts
    print "Answer? "
    response = ""
    while response.length < correct.length
      input = STDIN.getch.chomp.downcase.gsub(/,| /,"").chars.sort.join
      print input
      if input.match(/q/)
        print_results(no_quest,no_right,no_wrong)
        exit
      end
      if input.match(/[a-e]/)
        response = response+input
      end
    end
    puts
    puts
    answer = answer.wrap_text
    no_quest = no_quest + 1
    if response == correct
      no_right = no_right + 1
      puts answer.green
    else
      no_wrong = no_wrong + 1
      puts answer.red
    end
  end
  print_results(no_quest,no_right,no_wrong)
  return
end

# Process options

begin
  option = Long.getopts(
    [ "--list",    "-l", BOOLEAN ],  # List quizes
    [ "--random",  "-r", BOOLEAN ],  # Randomise quizes
    [ "--mix",     "-m", BOOLEAN ],  # Mix choices between questions
    [ "--quiz",    "-q", REQUIRED ], # Quiz
    [ "--help",    "-h", BOOLEAN ],  # Print help information
    [ "--version", "-V", BOOLEAN ]   # Print version information
  )
rescue
  print_usage()
  exit
end

if option["l"]
  list_quizes()
end

if option["h"]
  print_usage()
end

if option["V"]
  print_version()
end

# Ask questions in random order

if option ["r"]
  random = true
else
  random = false
end

# Mix choices between questions

if option ["m"]
  mix = true
else
  mix = false
end

if option["q"]
  handle_quiz(option["q"],random,mix)
end
